lib/GPUArraysCore/src/GPUArraysCore.jl

```diff
@@ -111,5 +111,13 @@ macro allowscalar(ex)
     end
 end
 
+"""
+    backend(T::Type)
+    backend(x)
+
+Gets the GPUArrays back-end responsible for managing arrays of type `T`.
+"""
+backend(::Type) = error("This object is not a GPU array") # COV_EXCL_LINE
```

Comment:
Now that we're touching this code anyway (and before anybody depends on it): Is this OK invalidation-wise? We don't want to invalidate all of StructArrays when actually importing a package that implements `GPUArraysCore.backend`.

Reply:
At present, we only want to extend GPU broadcast to `StructArrays` by dispatching with `AbstractGPUArrayStyle` and fusing the kernal function in `GPUArrays.jl`.
We will not write our own kernals. So invalidation (if it exits) would be limited to broadcast code.

BTW, I don't think it's a good idea to add more GPU support (e.g. `sum`, `mapreduce`...) in `StructArrays.jl`.
Type-based dispatch seems always a pain in this case, and we don't want to write our own kernal.
Something like `CUDA.sum` makes more sense to me.

Reply:
If you're using the broadcast kernel from GPUArrays.jl, you're depending on the full package, so why do you need `backend` in GPUArraysCore? I thought you would be calling `GPUArraysCore.backend` from StructArrays.

Reply:
There's no need for us to depend on the whole `GPUArrays.jl`. We'll define the following in `StructArrays.jl`
```julia
# Unwrapper to recover the behaviour defined by parent style.
@inline function Base.copyto!(dest::AbstractArray, bc::Broadcasted{StructArrayStyle{S, N}}) where {S, N}
    return copyto!(dest, convert(Broadcasted{S}, bc))
end
@inline function Broadcast.materialize!(::StructArrayStyle{S}, dest, bc::Broadcasted) where {S}
    return Broadcast.materialize!(S(), dest, bc)
end
```
If `S<:AbstractGPUArrayStyle` then we'll fallback to `GPUArrays._copyo!` (See https://github.com/JuliaGPU/GPUArrays.jl/pull/393)
And `AbstractGPUArrayStyle` would appear only if user load a GPU package.

Reply:
Then I fail to see why `backend` needs to be put in GPUArraysCore?

Reply:
_copyto! calls backend(dest) like all other kernels. The above dispatch will fail with “This object is not a gpu array” at present. The last piece we need is overloading backend. Just as the first take, we hope this function to live in a lightweight package, so…

Reply:
OK, you want to extend `backend`, that makes sense now. That's fine for now, but once Adapt has something like `get_computing_device` we might want to get rid of `backend` here.

Reply:
The invalidation concern still holds though: importing e.g. CUDA.jl which further extends the `backend` function may invalidate previously-inferred broadcast invocations (because it's so loosely typed here -- but I'm not sure).